url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/utility/knnwithcryop.java

```diff
@@ -0,0 +1,651 @@
+						filename = "./database/CA";
+					}else if(line.equalsIgnoreCase("c")){
+						System.out.println("Input the rate:");
+						String rate = in.nextLine();
+						filename = "./database/GO" + rate;
+						nbFileName = "./input/GO" + rate + ".pd.dat";
+					}else{
+						break;
+					}
+					System.out.println("begin loading " + filename + " ...");
+					MyRtree[] myrtrees = new MyRtree[1];
+					SecurityTree[] srtrees = new SecurityTree[1];
+					LoadMyRTree(filename, true, myrtrees, srtrees, false);
+					myrtree = myrtrees[0];
+					srtree = srtrees[0];
+					MyRtree[] myrtrees_kd = new MyRtree[1];
+					SecurityTree[] srtrees_kd = new SecurityTree[1];
+					LoadMyRTree(filename + "_kd", true, myrtrees_kd, srtrees_kd, true);
+					myrtree_kd = myrtrees_kd[0];
+					srtree_kd = srtrees_kd[0];
+					System.out.println("Load Successfully!");
+					VoronoiNeighbors vn = new VoronoiNeighbors();
+					vn.loadVoronoiNeighbors(nbFileName, nbHashMap);
+					break;
+				}
+			}else if(line.equalsIgnoreCase("d")){
+				IS_BATCH_QUERY = false;
+				System.out.println("===========start query===========");
+				while(true){
+					System.out.println("(a) new query\n(b) exit");
+					line = in.nextLine();
+					if(line.equalsIgnoreCase("a")){
+						System.out.println("Please input the k of knn:");
+						int kNum = Integer.parseInt(in.nextLine());
+						System.out.println("Please input the coordinate of point:");
+						String[] coords = in.nextLine().split(" ");
+						System.out.println("Please input the limit:");
+						int limit = Integer.parseInt(in.nextLine());
+						System.out.println("Please input whether to use parallel (y/n)");
+						boolean isParallel = in.nextLine().equalsIgnoreCase("y");
+						Point pt = new Point(new double[]{Double.parseDouble(coords[0]), Double.parseDouble(coords[1]), 0});
+						query(kNum, pt, new PrintWriter(System.out), null, -1, limit, isParallel);				
+					}else{
+						System.out.println("===========end query===========");
+						break;
+					}
+				}
+			}else if(line.equalsIgnoreCase("e")){
+				IS_BATCH_QUERY = true;
+				System.out.println("Please input query's filename");
+				String filename = in.nextLine();
+				System.out.println("Please input answer of query's file name");
+				String ans_file_name = in.nextLine();
+				System.out.println("Please input the limit:");
+				int limit = Integer.parseInt(in.nextLine());
+				System.out.println("Please input whether to use parallel (y/n)");
+				boolean isParallel = in.nextLine().equalsIgnoreCase("y");
+				StatisticForAuth[] stat = new StatisticForAuth[6];
+				//int[] k_of_knn = {1, 2, 4, 8, 16, 32, 64, 128};
+//				int[] k_of_knn = {128, 1, 64, 32, 16, 8, 4, 2};
+				int[] k_of_knn = {128, 128, 1, 1, 64, 64, 16, 8, 4, 2};
+				for(int i = 0; i < 8; i ++){
+					System.out.println("k = " + k_of_knn[i]);
+					final PrintWriter datapw = new PrintWriter(new FileOutputStream(new File(ans_file_name + "_" + new Integer(k_of_knn[i]).toString() + ".data")));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *datapw*. The resource is referenced at lines 220, 234, 247. The resource is closed at line 248. There are other execution paths that do not contain closure statements, for example, when Double.parseDouble throws an exception. Ensure *datapw* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).